### PR TITLE
bg/meson-paths

### PR DESCRIPTION
**Pull request by bengal**
Some patches to fix the configuration of build paths with meson.

See https://gitlab.freedesktop.org/NetworkManager/NetworkManager/issues/19

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -885,15 +885,6 @@ config_extra_h = configuration_data()
 
 config_extra_h.set_quoted('BINDIR',            nm_bindir)
 config_extra_h.set_quoted('DATADIR',           nm_datadir)
-if enable_dhclient
```
Comment (thom311) [inline meson.build]: is this patch right? config-extra.h.meson still contains `mesondefine DHCPCANON_PATH`. Also, don't you need to set_quoted() now somewhere else? Is this fixed by the following commit? If so, it seems this patch is broken and should be merged with the following commit.

File: meson.build [hunk 2]
```diff
@@ -505,11 +505,14 @@ enable_ppp = get_option('ppp')
...
   assert(cc.has_header('pppd/pppd.h'), 'couldn\'t find pppd.h. pppd development headers are required')
 
-  locations = get_option('pppd')
-  pppd = find_program(locations, required: false)
-  assert(pppd.found(), 'pppd required but not found, please provide a valid pppd path or use -Dppp=false to disable it')
+  pppd_path = get_option('pppd')
+  if pppd_path == ''
+    pppd = find_program('pppd', '/sbin/pppd', '/usr/sbin/pppd', required: false)
+    assert(pppd.found(), 'pppd required but not found, please provide a valid pppd path or use -Dppp=false to disable it')
+    pppd_path = pppd.path()
```
[the remark this answers is not in the document]
Comment (bengal) [reply meson.build]: > the commit message says, one can specify a path that does not exist. Still, there is an `assert(pppd.found())` here.

Yes, but the assertion is reached only when `pppd_path == ''`, i.e. when autodetecting the path.

File: meson.build [hunk 3]
```diff
@@ -652,15 +652,34 @@ endif
...
+  path = get_option(prog[0])
+  if path == ''
+    search_paths = [ prog[0] ]
+    foreach path : prog[1]
+      search_paths += (path + '/' + prog[0])
+    endforeach
+    exe = find_program(search_paths, required : false)
+    path = exe.found() ? exe.path() : prog[2]
+  endif
+  name = prog.length() > 3 ? prog[3] : (prog[0].to_upper() + '_PATH')
```
[the remark this answers is not in the document]
Comment (bengal) [reply meson.build]: Ok, I added a commit to rename the path and remove the hack.

File: meson.build (hunk 1 above)
Comment (bengal) [reply meson.build]: I removed the leftover defines from config-extra.h.meson now.

The `config_h.set_quoted('DHCPCANON_PATH', dhcpcanon.path())` is already there (previously the path was added to both config.h and config-extra.h)

**Comment by thom311** [comment]
lgtm

**Comment by bengal** [comment]
Merged as 4a1ff8ad8c543a76b2430a79fb8ffee52e8d3f56

**Pull request closed without merging**